kong/plugins/zipkin/handler.lua

```diff
@@ -1,23 +1,144 @@
-local new_tracer = require "opentracing.tracer".new
-local zipkin_codec = require "kong.plugins.zipkin.codec"
-local new_random_sampler = require "kong.plugins.zipkin.random_sampler".new
 local new_zipkin_reporter = require "kong.plugins.zipkin.reporter".new
-local OpenTracingHandler = require "kong.plugins.zipkin.opentracing"
+local new_span = require "kong.plugins.zipkin.span".new
+local to_hex = require "resty.string".to_hex
 
--- Zipkin plugin derives from general opentracing one
-local ZipkinLogHandler = OpenTracingHandler:extend()
-ZipkinLogHandler.VERSION = "0.2.1"
+local subsystem = ngx.config.subsystem
+local fmt = string.format
 
+local ZipkinLogHandler = {
+  VERSION = "0.2.1",
+  -- We want to run first so that timestamps taken are at start of the phase
+  -- also so that other plugins might be able to use our structures
+  PRIORITY = 100000,
+}
 
-function ZipkinLogHandler.new_tracer(conf)
-  local tracer = new_tracer(new_zipkin_reporter(conf), new_random_sampler(conf))
-  tracer:register_injector("http_headers", zipkin_codec.new_injector())
-  tracer:register_extractor("http_headers", zipkin_codec.new_extractor(kong.log.warn))
-  return tracer
+local reporter_cache = setmetatable({}, { __mode = "k" })
+
+local math_random = math.random
+
+local baggage_mt = {
+  __newindex = function()
+    error("attempt to set immutable baggage")
+  end,
+}
+
+
+local function hex_to_char(c)
+  return string.char(tonumber(c, 16))
```

Comment:
This is potentially hot path, therefore `string.char` and `tonumber` should be cached.